### PR TITLE
Proposal for streams changes listener.

### PR DESCRIPTION
**Pull request by jcjimenez**
empty_stream

### CODE REVIEW

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala [hunk 1]
```diff
@@ -61,8 +63,13 @@ object ProjectFortis extends App {
...
 
-    val sparksession = SparkSession.builder().config(conf).getOrCreate()
-    val ssc = new StreamingContext(sparksession.sparkContext, batchDuration)
+    val sparkContext = SparkContext.getOrCreate(conf)
+    val ssc = new StreamingContext(sparkContext, batchDuration)
+    ssc.checkpoint(fortisSettings.progressDir)
+    ssc
+  }
+
+  private def attachToContext(ssc:StreamingContext): Boolean = {
```
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala:71]: I factored out `attachToContext` from `createStreamingContext` in order to call it in the loop below.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        logInfo(s"Service Bus message for updated streams received; Re-scheduling streaming context stop for ${contextStopWaitTimeMillis} milliseconds from now.")
+        this.timer.get.cancel()
+      } else {
+        logInfo(s"Service Bus message for updated streams received; Requesting streaming context stop in ${contextStopWaitTimeMillis} milliseconds.")
+      }
+
+      val t = new Timer()
+      t.schedule(new TimerTask {
+        override def run(): Unit = {
+          ssc.stop(stopSparkContext = true, stopGracefully = true)
```
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala]: Not sure if just calling `stop` is enough to dispose of the open streams.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala [hunk 3]
```diff
@@ -6,6 +6,8 @@ object Constants {
   val SparkStreamingBatchSizeDefault = 5
 
   val EventHubProgressDir = "eventhubProgress"
+  val ContextStopWaitTimeMillis = 5000
```
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala]: I think I need to change this to allow for more wiggle room in cases where the user is adding more than one stream. Maybe increase this to 1 minute?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala [hunk 4]
```diff
@@ -18,5 +20,7 @@ object Constants {
     val ManagementBusConnectionString = "FORTIS_SB_CONN_STR"
     val ManagementBusConfigQueueName = "FORTIS_SB_CONFIG_QUEUE"
     val ManagementBusCommandQueueName = "FORTIS_SB_COMMAND_QUEUE"
+    val ContextStopWaitTimeMillis = "FORTIS_CONTEXT_STOP_WAIT_TIME_MILLIS"
```
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala]: nit: Can you make these name a little shorter?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala [hunk 5]
```diff
@@ -85,14 +92,28 @@ object ProjectFortis extends App {
...
+    true
   }
 
   // Main starts here
-  val ssc = StreamingContext.getOrCreate(fortisSettings.progressDir, createStreamingContext)
+  while(true) {
+    logInfo("Creating streaming context.")
+    val ssc = StreamingContext.getOrCreate(fortisSettings.progressDir, createStreamingContext)
+    var didAttach = attachToContext(ssc)
+    while (!didAttach) {
```
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala]: Do you need to store `didAttach`?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.{Timer, TimerTask}
+
+import com.microsoft.azure.servicebus._
+import com.microsoft.azure.servicebus.primitives.ConnectionStringBuilder
+import com.microsoft.partnercatalyst.fortis.spark.logging.Loggable
+import org.apache.spark.streaming.StreamingContext
+
+object StreamsChangeListener {
+
+  @volatile @transient private var queueClient: QueueClient = _
```
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala]: This object won't be serialized so you should be able to make this a `val` right? In that way, it should also be safe to access from the inner class helper thread.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    override def onMessageAsync(message: IMessage): CompletableFuture[Void] = {
+      if (message.getLabel != "streamsDidChange") {
+        return CompletableFuture.completedFuture(null)
+      }
+
+      val contextStopWaitTimeMillis = settings.contextStopWaitTimeMillis
+
+      if (this.timer.isDefined) {
```
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/StreamsChangeListener.scala]: How about a `match` here? :)

**Review by kevinhartman — APPROVED** [review]

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala (hunk 5 above)
Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/ProjectFortis.scala]: Good point!

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala (hunk 4 above)
Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala]: lol, sure

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala (hunk 3 above)
Comment (kevinhartman) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Constants.scala]: Shouldn't the user need to click an `apply` button before we issue the restart command?

**Comment by kevinhartman** [comment]
Also if you haven't already, make sure to delete the contents of the old checkpoint directory prior to restarting streaming.

**Comment by kevinhartman** [comment]
@jcjimenez I'd also like to request that we delay merging this feature until we've stabilized a bit since making the project more autonomous could add hiccups to the deploy, test, re-deploy loop.

**Comment by jcjimenez** [comment]
OK, I've reduced the scope of this PR to just waiting for `fortis.streams` records to be present.

**Pull request merged**